### DISCUSSION TITLE
Does job creation wait until ActiveRecord transaction commits?

### DISCUSSION THREAD

**Discussion by maxim** [bensheldon/good_job#1058] [Q&A] [answered]
In this situation:

```ruby
ApplicationRecord.transaction do
  user.update! job_pending: true
  Job.perform_later
  sleep 20
end
```

Will the job get run before the transaction commits, and therefore not find the pending user?

**Comment by bensheldon** [comment, marked as answer]
GoodJob uses Active Record so it will share the same behavior as Active Record. The job will be enqueued (inserted/committed) when the outer transaction commits (unless you're using Active Record's multiple databases feature and GoodJob is in a different database than the model; but that's unlikely). 

I had the thought that when using `:inline` execution mode it will still work too because the job will be performed before the transaction closes, but it will be performed within the transaction ("inline") which will have the updated data). 

So yes, that will all work properly 👍

**Reply by maxim** [reply]
@bensheldon wow, thanks for the quick and helpful response. 🎉

> when using `:inline` execution mode it will still work

I believe, this can vary depending on the order. For example

```ruby
ApplicationRecord.transaction do
  Job.perform_later
  user.update! job_pending: true
end
```

This might work in regular mode, but not work inline, right?

**Reply by bensheldon** [reply]
Correct, when running `:inline` order does matter because it's within the transaction. But otherwise, no. Though it'd probably still be bad/confusing to write it that way 😆
